File: Data Structures & Algorithms/binary-search/submission-0.cpp
/**
nums = [-1,0,2,4,6,8], target = 2
m = 3 -> r = 2
m = 1 -> l = 2

*/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l=0, r=nums.size()-1;

        while (l < r) {
            int m = (r+l)/2;

            if (target == nums[m]) {
                return m;
            } else if (target < nums[m]) {
                r = m-1;
            } else {
                l = m+1;
            }
        }

        if (target == nums[l]) return l;
        if (target == nums[r]) return r;

        return -1;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        # given list of distinct integers
        # given target
        # list all unique combinations of nums that sum to target
        # can reuse a number unlimited times

        """
        APPROACH
        - backtracking
        - go through each coin and add them to a growing list
        - if total == target, add that to output array
        - if total > target, backtrack
        - to avoid duplications, only consider coins that are equal or of a later position in the list of coins
        """
        
        res = []

        def backtrack(stack: List[int], start_idx: int, total: int) -> None:
            nonlocal res
            if start_idx >= len(nums):
                return
            if total == target:
                res.append(stack)
                return
            if total > target:
                return
            
            for i in range(len(nums)):
                if i >= start_idx:
                    stack.append(nums[i])
                    total += nums[i]
                    backtrack(list(stack), i, total)
                    stack.pop()
                    total -= nums[i]
            
        backtrack([], 0, 0)

        # return: combinations that sum to order
        return res
